sui_core/src/authority/authority_store.rs

```diff
@@ -753,61 +692,55 @@ impl<
         // certs which may overwrite newer objects with older ones.  This can be removed once we have
         // an object storage supporting multiple object versions at once, then there is idempotency and
         // old writes would be OK.
-        {
-            // Acquire the lock to ensure no one else writes when we are in here.
-            let _mutexes = self.acquire_locks(&owned_inputs[..]).await;
-
-            // NOTE: We just check here that locks exist, not that they are locked to a specific TX.  Why?
-            // 1. Lock existence prevents re-execution of old certs when objects have been upgraded
-            // 2. Not all validators lock, just 2f+1, so transaction should proceed regardless
-            //    (But the lock should exist which means previous transactions finished)
-            // 3. Equivocation possible (different TX) but as long as 2f+1 approves current TX its fine
-            if USE_LOCKS {
-                self.lock_service.locks_exist(owned_inputs.clone()).await?;
-            }
 
-            if let Some(next_seq) = seq_opt {
-                // Now we are sure we are going to execute, add to the sequence
-                // number and insert into authority sequence.
-                //
-                // NOTE: it is possible that we commit to the database transactions
-                //       out of order with respect to their sequence number. It is also
-                //       possible for the authority to crash without committing the
-                //       full sequence, and the batching logic needs to deal with this.
-                write_batch = write_batch.insert_batch(
-                    &self.executed_sequence,
-                    std::iter::once((next_seq, transaction_digest)),
-                )?;
-            }
+        // NOTE: We just check here that locks exist, not that they are locked to a specific TX.  Why?
+        // 1. Lock existence prevents re-execution of old certs when objects have been upgraded
+        // 2. Not all validators lock, just 2f+1, so transaction should proceed regardless
+        //    (But the lock should exist which means previous transactions finished)
+        // 3. Equivocation possible (different TX) but as long as 2f+1 approves current TX its fine
+        if USE_LOCKS {
+            self.lock_service.locks_exist(owned_inputs.clone()).await?;
+        }
 
-            // Atomic write of all data other than locks
-            write_batch.write()?;
-            trace!("Finished writing batch");
+        if let Some(next_seq) = seq_opt {
```

Comment:
Is there no "time of check / time of use" problem here: consider two instances of the same cert going through execution. They both reach the locks check on 701. They both pass at the same time. And now we sequence on 705 the same transaction twice (with different `next_seq`!). 

This was a/the purpose of the lock before: to ensure the check (line 702) and the write (line 720) are atomic. So that the second attempt at writing the same tx fails, and returns either Err or the result of the first.

Reply:
You are right that the problem I outline is not solved by mutexes alone but by checking for the object locks are present.
For reference the code used to look like this:
https://github.com/MystenLabs/sui/blob/3f627838c4b0a2fd329c11cea77919295f8a0d62/sui_core/src/authority/authority_store.rs#L754-L788
and now like this:
https://github.com/MystenLabs/sui/blob/9ce05fbf45906e6c2d48b7c9a4f6039733fee742/crates/sui-core/src/authority/authority_store.rs#L773-L828

Note the following:
* We lock the mutexes on line 775.
* The we check for the object locks on line 783 (on the second attempt to execute this will return an Err).
* Then write the rest and delete the locks on line 824.
* Then unlock

So the check for object locks + delete locks on first execution is atomic. And that ensures we execute once, and we only sequence once.

The problem with having to remember the transaction digest for ever is that this state grows for ever, as compared with the state of the locks which tracks live objects and the locks we maintain.

Reply:
@gdanezis isn't the root problem here that the tx_seq that is issued is different?  If we allow the concurrent execution of the same TX, but also allow a different tx_seq to be issued, that breaks the principle of idempotency.  

To truly get rid of mutexes we need truly idempotent execution and re-execution.   Using mutexes and locks is kind of a poor substitute for having that, or we should just truly try to prevent concurrent or repeat execution.  

We already check that previous TXes are executed by looking up a certificate (which does require keeping lots of TXes around already).   I don't think a TX digest based lock would really operate any differently.

There are other solutions too, such as keeping a list of already executing transactions, which I had proposed earlier.

Reply:
Also, by the time `batch_update_objects()` is called, the `seq_opt` has already been issued.  While the mutex prevents re-writing the transaction, the whole principle of this change is that it should be OK to re-write everything a second time.   Why not just re-issue the same TX sequence number?

Reply:
Actually @gdanezis I think I have a solution to the whole thing.   Really, `executed_sequence` and the sequencing should be done differently, as part of a `CertificateStore`.   In the medium term the sequence number logic that we are debating here will be removed.  Thus I don't think this will even be an issue going forward.    It will be the job of a certificate/transaction store to issue sequence numbers, as transactions form part of the log output of Sui.   

So the question is how much do we care to have this corner case be taken cared of in the next say 2-3 weeks.  If not very much, add an issue here, assign it to me, and let's move on.  (And if you do, comment in the issue and I'll think of something short term)